include/region_map.h

```diff
@@ -1,142 +1,1414 @@
+		case 2: get_map<2>().update_box(box_cast<2>(box), value); break;
+		case 3: get_map<3>().update_box(box_cast<3>(box), value); break;
+		default: assert(false);
+		}
+	}
+
+	/**
+	 * Returns all entries in the region map that intersect with the request region.
+	 *
+	 * @returns A list of boxes clamped to the request region, and their associated values.
+	 */
+	std::vector<std::pair<GridBox<3>, ValueType>> get_region_values(const GridRegion<3>& request) const {
+		region_map_detail::assert_dimensionality(request, m_dims);
+		std::vector<std::pair<GridBox<3>, ValueType>> results;
+		request.scanByBoxes([&](const GridBox<3>& box) {
+			const auto r = get_region_values(box);
+			results.insert(results.begin(), r.cbegin(), r.cend());
+		});
+		return results;
+	}
+
+	/**
+	 * Returns all entries in the region map that intersect with the request box.
+	 *
+	 * @returns A list of boxes clamped to the request box, and their associated values.
+	 */
+	std::vector<std::pair<GridBox<3>, ValueType>> get_region_values(const GridBox<3>& request) const {
+		using namespace region_map_detail;
+		std::vector<std::pair<GridBox<3>, ValueType>> results;
+		switch(m_dims) {
+		// TODO: AllScale box doesn't support 0 dimensions, fall back to 1
+		case 0: {
+			const auto results1 = get_map<0>().get_region_values(box_cast<1>(request));
+			results.reserve(results1.size());
+			std::transform(
+			    results1.cbegin(), results1.cend(), std::back_inserter(results), [](const auto& p) { return std::make_pair(box_cast<3>(p.first), p.second); });
+		} break;
+		case 1: {
+			const auto results1 = get_map<1>().get_region_values(box_cast<1>(request));
+			results.reserve(results1.size());
+			std::transform(
+			    results1.cbegin(), results1.cend(), std::back_inserter(results), [](const auto& p) { return std::make_pair(box_cast<3>(p.first), p.second); });
+		} break;
+		case 2: {
+			const auto results2 = get_map<2>().get_region_values(box_cast<2>(request));
+			results.reserve(results2.size());
+			std::transform(
+			    results2.cbegin(), results2.cend(), std::back_inserter(results), [](const auto& p) { return std::make_pair(box_cast<3>(p.first), p.second); });
+		} break;
+		case 3: {
+			results = get_map<3>().get_region_values(box_cast<3>(request));
+		} break;
+		default: assert(false);
+		}
+		return results;
+	}
+
+	/**
+	 * Applies a function f to every value within the region map and stores the result in its place.
+	 */
+	template <typename Functor>
+	void apply_to_values(const Functor& f) {
+		static_assert(std::is_invocable_r_v<ValueType, Functor, const ValueType&>, "Functor must receive and return a value of type ValueType");
+
+		switch(m_dims) {
+		case 0: get_map<0>().apply_to_values(f); break;
+		case 1: get_map<1>().apply_to_values(f); break;
+		case 2: get_map<2>().apply_to_values(f); break;
+		case 3: get_map<3>().apply_to_values(f); break;
+		default: assert(false);
+		}
+	}
+
+  private:
+	int m_dims;
```

Comment:
There's a lot of DRY potential here: `m_dims` duplicates the variant index, and all except the constructor switches can be replaced by a `std::visit` or `utils::match`. I gave this a try and it is currently complicated by the missing `GridBox<0>`, so I agree to merge this - but we should remember to refactor it as soon as we have a custom box implementation.

Reply:
Right, using `std::visit` would be neat. Somehow I always keep forgetting about that...